torch/autograd/_functions/reduce.py

```diff
@@ -71,7 +71,10 @@ def safe_zeros_backward(inp, dim):
             exclusive_normal = exclusive_normal_nocp.cumprod(dim)
 
             def reverse_dim(var, dim):
-                return var.index_select(dim, Variable(torch.arange(var.size(dim) - 1, -1, -1)).long())
+                index = Variable(torch.arange(var.size(dim) - 1, -1, -1)).long()
+                if var.is_cuda:
+                    index = index.cuda()
```

Comment:
This only works if your current device is the device you want to send the `index` to.
I think a better solution would be to replace this by something like
```python
idx = torch.arange(var.size(dim) -1, -1, -1, out=var.data.new().long())
return var.index_select(dim, Variable(idx))
```